core/model-api/src/main/java/org/eclipse/rdf4j/model/base/AbstractIRI.java

```diff
@@ -34,17 +86,92 @@ public String stringValue() {
 	@Override
 	public boolean equals(Object o) {
 		return this == o || o instanceof IRI
-				&& Objects.equals(toString(), ((Value) o).toString()); // !!! use stringValue()
+				&& toString().equals(o.toString()); // !!! use stringValue()
 	}
 
 	@Override
 	public int hashCode() {
-		return Objects.hashCode(toString()); // !!! use stringValue()
+		return toString().hashCode(); // !!! use stringValue()
 	}
 
 	@Override
 	public String toString() {
 		return stringValue();
 	}
 
+	////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+	private static class UnaryIRI extends AbstractIRI {
```

Comment:
I think this may be over-optimising things somewhat. Keep in mind that most database implementations will have their own ValueFactory (and therefore own IRI) implementations.

Reply:
More of a matter of convenience, as the merging/splitting logic is reversed in the two cases, but I'll review the matter.